File: tokenizer.cpp
#include <stdint.h>
#include <ctype.h>
#include "tokenizer.h"

static inline int prefix(const char *pre, const char *str)
	{
	while (*pre)
		if (*pre++ != *str++)
			return 0;
	return 1;
	}

static inline char lower(char c)
	{
	if (c < 'a')
		c += 'a' - 'A';

	return c;
	}

static inline char upper(char c)
	{
	if ('Z' < c)
		c -= 'a' - 'A';
	
	return c;
	}

tokenizer::tokenizer(char *doc, size_t len)
	{
	document = doc;
	length = len;
	index = 0;
	}

enum token_type tokenizer::next(char *buffer)
	{
	for (;;)
		{
		// Whitespace
		while(index < length && isspace(document[index]))
			index++;
		// EOF
		if (index == length)
			break;
		// Doc ID
		else if (prefix("<DOCNO>", &document[index]))
			{
			index += sizeof("<DOCNO>");

			while(index < length && isspace(document[index]))
				index++;

			int i = 0;
			while (i < 256 && i + index < length && document[index + i] != '<' && !isspace(document[index + i]))
				{
				buffer[i+4] = document[index + i];
				i++;
				}
			((uint32_t *)buffer)[0] = i;
			buffer[i+4] = '\0';

			index += i;

			return DOCNO;
			}
		// Ignored tags
		else if (document[index] == '<')
			{
			index++;
			while (index < length && document[index] != '>')
				index++;
			index++;
			continue;
			}
		// Number
		else if (isdigit(document[index]))
			{
			int i = 0;
			while (i < 256 && i + index < length && isdigit(document[index + i]))
				{
				buffer[i+4] = document[index + i];
				i++;
				}
			((uint32_t *)buffer)[0] = i;
			buffer[i+4] = '\0';

			index += i;

			return WORD;
			}
		// Word
		else if (isalpha(document[index]))
			{
			int i = 0;
			while (i < 256 && i + index < length && isalpha(document[index + i]))
				{
				buffer[i+4] = upper(document[index + i]);
				i++;
				}
			((uint32_t *)buffer)[0] = i;
			buffer[i+4] = '\0';

			index += i;

			return WORD;
			}
		// Something else we don't want
		else
			index++;
		}

	return END;
	}

File: tokenizer.zig
//	TOKENIZER.ZIG
//	-------------
//	Copyright (c) Vaughan Kitchen
//	Released under the ISC license (https://opensource.org/licenses/ISC)

const std = @import("std");

pub const Token = struct {
    pub const Type = enum { docno, word, eof };

    token: []const u8,
    type: Type,
};

pub const Tokenizer = struct {
    index: usize,
    doc: []const u8,

    pub fn init(doc: []const u8) Tokenizer {
        return .{ .index = 0, .doc = doc };
    }

    pub fn reinit(t: *Tokenizer, doc: []const u8) void {
        t.index = 0;
        t.doc = doc;
    }

    pub fn next(t: *Tokenizer) Token {
        while (true) {
            // Whitespace
            while (t.index < t.doc.len and std.ascii.isWhitespace(t.doc[t.index])) {
                t.index += 1;
            }
            // EOF
            if (t.index == t.doc.len) {
                break;
            } else if (std.mem.startsWith(u8, t.doc[t.index..], "<DOCNO>")) {
                t.index += std.mem.len("<DOCNO>");

                while (t.index < t.doc.len and std.ascii.isWhitespace(t.doc[t.index]))
                    t.index += 1;

                var i: usize = 0;
                while (i < 256 and i + t.index < t.doc.len and t.doc[t.index + i] != '<' and !std.ascii.isWhitespace(t.doc[t.index + i])) {
                    i += 1;
                }

                const out = Token{
                    .token = t.doc[t.index .. t.index + i],
                    .type = Token.Type.docno,
                };

                t.index += i;

                return out;
            }
            // Ignored tags
            else if (t.doc[t.index] == '<') {
                t.index += 1;
                while (t.index < t.doc.len and t.doc[t.index] != '>')
                    t.index += 1;
                t.index += 1;
                continue;
            }
            // Number
            else if (std.ascii.isDigit(t.doc[t.index])) {
                var i: usize = 0;
                while (i < 256 and i + t.index < t.doc.len and std.ascii.isDigit(t.doc[t.index + i])) {
                    i += 1;
                }

                const out = Token{
                    .token = t.doc[t.index .. t.index + i],
                    .type = Token.Type.word,
                };

                t.index += i;

                return out;
            }
            // Word
            else if (std.ascii.isAlpha(t.doc[t.index])) {
                var i: usize = 0;
                while (i < 256 and i + t.index < t.doc.len and std.ascii.isAlpha(t.doc[t.index + i])) {
                    // buffer[i] = char_tolower(t.doc[t.index + i]);
                    i += 1;
                }

                const out = Token{
                    .token = t.doc[t.index .. t.index + i],
                    .type = Token.Type.word,
                };

                t.index += i;

                return out;
            }
            // Something else we don't want
            else {
                t.index += 1;
            }
        }
        const out = Token{
            .token = t.doc[0..0],
            .type = Token.Type.eof,
        };
        return out;
    }
};